### DISCUSSION TITLE
Could we adjust the PRESENCE_DETECTION_LOOP_DELAY programmatically ?

### DISCUSSION THREAD

**Discussion by nicolinuxfr** [tesla-local-control/tesla_ble_mqtt_docker#62] [Ideas]
I'm beginning to play with the project to automated my garage door based on the presence or absence of my car. The 120 seconds fixed delay between checks of presence is not ideal in this situation : it's short, maybe too short, when the car sits in the garage for  the night or is away during the day. But when I am in the car ready to go, these 120 seconds are too long to reliably know if the car is still in the garage or out. 

Hence my idea : could it be possible to adjust the delay through Home Assistant ? I could automate to make it longer when the car is asleep and shorter when it wakes up or when I come back home.

**Comment by iainbullock** [comment]
It would be possible, though probably not something we are likely to be working on any time soon. 

If you are up for modifying the code yourself, you could do it. 

In particular look at mqtt-discovery.sh and mqtt-listen.sh to see how we setup entities that are not car commands but cause the code to do something e.g. function setupDeployKeyButton() in mqtt-discovery.sh creates a button entity,  listen_to_mqtt() in mqtt-listen.sh responds to the button being pressed at this option:

```
        deploy-key)
          log_debug "deploy-key; calling deployKeyMain()"
          deployKeyMain $vin
          ;;

```

So you could create a new entity in mqtt-discovery.sh and have a corresponding response in mqtt-listen.sh which would change PRESENCE_DETECTION_LOOP_DELAY

Quite a bit of work but it can be done. I can help if you want to have a go

**Reply by nicolinuxfr** [reply]
Thanks for your answer and suggestion !

In the meantime, I found out that Home Assistant itself can "see" the car's BLE and follow it, at a much quicker pace it seems. So maybe I'll use HA for this task.

And I can't manage to deploy the key to control the car using this project. I'm guessing my Bluetooth USB key is too far, though it's really close to the car. I'm not sure if it's a viable project for me, I will try to deploy the key while the computer I'm using is inside the cabin, but it won't be a viable option on a daily basis.

**Reply by iainbullock** [reply]
Good to have a solution for your presence application. If you can't get the key to deploy and the Bluetooth adapter is close it is most likely that Tesla doesn't like your Bluetooth adapter. Anecdotally we believe it has to support Bluetooth 5.0 or greater. What are you using?

**Reply by nicolinuxfr** [reply]
I'm using a Sabrent ‎BT-UB40 which was on the recommended list of HA, but I am just realizing it's BT 4.0 only. I'm guessing it explains why it's not working then…

**Reply by iainbullock** [reply]
Yes I think that's it. If you are looking for a dedicated device for this application which you can locate close to your car I recommend a Raspberry Pi with built in Bluetooth. I use a RPi3b+ which works well

**Reply by nicolinuxfr** [reply]
I was thinking about that option, but I have a full Mini PC right next to the car. I may try Bluetooth 5 but the Home Assistant Bluetooth documentation does not list any, so it's a little harder to know what to order…

The RPi3b+ is only Bluetooth 4 though, shouldn't it *not* work as well ?

**Reply by iainbullock** [reply]
I hadn't realised but you are right RPi3b+ Bluetooth is 4.2. So 5.0 isn't the minimum! 

We are finding people are having varying success depending on the Bluetooth adapter used. The Tesla libraries that this project uses is fussy but doesn't document what works and what doesn't. 

What error messages are you getting from the Docker logs when you try to deploy?

**Reply by nicolinuxfr** [reply]
> What error messages are you getting from the Docker logs when you try to deploy?

Messages like this one : 

```
tesla_ble_mqtt  | teslaCtrlSendKey; Retrying in 5 seconds
tesla_ble_mqtt  | Attempt 2/5 to delivery the public key to vin XXXXXX
tesla_ble_mqtt  | teslaCtrlSendKey; Error: context deadline exceeded
tesla_ble_mqtt  | Could not send the key; Is the car awake and sufficiently close to the bluetooth adapter?
```

I do have a Raspberry Pi 3B+ laying around so I could try with it. My USB key seems to be BT 4.0, so maybe 4.2 is the minimum ? I need some time to set everything up on the the Pi, but I was wondering if I should try first to get my miniPC and BT key really close just for the key process.

**Reply by nicolinuxfr** [reply]
Well I forgot I had a Pi3 almost ready so the time needed to deploy the Docker image and… I have made progress !

```
tesla_ble_mqtt  | Attempt 1/5 to delivery the public key to vin XXXXXX
tesla_ble_mqtt  | KEY DELIVERED; IN YOUR CAR, CHECK THE CAR's CENTRAL SCREEN AND ACCEPT THE KEY USING YOUR NFC CARD
tesla_ble_mqtt  | acceptKeyConfirmationLoop; check if key was accepted by sending a ping command vin XXXXXX
tesla_ble_mqtt  | Attempt 1/5 sending Ping vehicule to vin:XXXXXX command:ping
tesla_ble_mqtt  | tesla-control send command:ping to vin:XXXXXX failed exit status 1.
tesla_ble_mqtt  | teslaCtrlSendCommand; Error: vehicle rejected request: your public key has not been paired with the vehicle
tesla_ble_mqtt  | teslaCtrlSendCommand; Retrying in 5 seconds
```

Unfortunately, it seems the procedure did not work. I was not aware I had to be in the car, but there was no message and I am not sure what I should do now ?

**Comment by iainbullock** [comment]
Did you complete the process with the NFC card in the car to accept the key? Does it show up as a key in the car's main screen (Locks->Keys), like mine below:

![IMG_5673-1](https://github.com/user-attachments/assets/11c443a3-50c6-4bae-8b71-38f3bdace9fa)

Note I renamed the key to TESLA_BLE_MQTT later, so yours won't be called that. Many people find it difficult to perform this step as the car doesn't always pop up the dialog to accept the key.
